### DISCUSSION TITLE
@angular-three VR Examples

### DISCUSSION THREAD

**Discussion by IRobot1** [nartc/angular-three#131] [Show and tell]
If you need examples of doing VR with @angular-three, see this [repo](https://github.com/IRobot1/ng3-vr-examples) and [demo](https://ng3vr.z9.web.core.windows.net/)

![image](https://user-images.githubusercontent.com/25032599/174513276-63d94f9e-4bca-4172-bacf-7a34b7d7f209.png)

Some of the functionality will eventually be moved into reusable @angular-three/webxr library.

**Comment by IRobot1** [comment]
Added [ng3-webxr](https://www.npmjs.com/package/ng3-webxr) for easily adding Web VR or AR functionality to an NGT project.

**Comment by IRobot1** [comment]
Updated [VR Demo](https://ng3vr.z9.web.core.windows.net) to include example of movable cameras and texture that shows the output of a camera.  
![image](https://user-images.githubusercontent.com/25032599/177672025-7d140f96-fdd4-4c03-be4e-58404608ae4f.png)

Use left controller to grab and move the cameras around.  
Use right controller to teleport around the room.

**Comment by IRobot1** [comment]
Updated [VR Demo](https://ng3vr.z9.web.core.windows.net/) to include paint example.  

![image](https://user-images.githubusercontent.com/25032599/179428909-e9fb8362-e5f7-4ece-b5f7-cde8d8ccc09c.png)

Pull trigger to paint
Joystick up and down to increase or decrease brush tip size

`paintbrush` is a directive that can be added to vr-controller and dynamically enabled or disabled.

**Comment by IRobot1** [comment]
Updated [VR Demo](https://ng3vr.z9.web.core.windows.net/) to include HTML GUI example.

![image](https://user-images.githubusercontent.com/25032599/179652815-4675aed8-53c1-4cb4-9983-993a5e379ed2.png)

Use pointer to interact with geometry properties inside a HTML Mesh panel.

**Comment by IRobot1** [comment]
Updated [VR Demo](https://ng3vr.z9.web.core.windows.net/) to include model scaling example.
![image](https://user-images.githubusercontent.com/25032599/180625899-23308c2c-6ade-4f3b-ac43-84c300144260.png)

Pull trigger on left and right controllers to start scaling model.
Move controllers closer to reduce scale.
Move controllers apart to increase scale.

Note the model does not appear correctly when visiting the room a second time.  Bug report pending.

**Comment by nartc** [comment]
@IRobot1 Hey, I have just released a beta version for v6, 6.0.0.beta-0, to add support for Angular 14 (14.1) and THREE 0.142.0

I don't have VR so can you try it out on VRs?

**Comment by IRobot1** [comment]
Sure can

**Reply by nartc** [reply]
Make sure to upgrade Angular to 14.1

**Comment by IRobot1** [comment]
Updated [VR Demo](https://ng3vr.z9.web.core.windows.net/) to include interactive buttons
![image](https://user-images.githubusercontent.com/25032599/183309362-2a4bffc9-22e1-42db-9de2-ab3937701584.png)

Buttons highlight/unhighlight on mouse or pointer hover.
Click or pull trigger to click button.
Animation is an optional directive that can be enabled on button.
Button can contain optional text and scaled as needed.

**Comment by IRobot1** [comment]
Updated [VR Demo](https://ng3vr.z9.web.core.windows.net/) to include interactive wall
![image](https://user-images.githubusercontent.com/25032599/183309626-41628eb6-9897-4a13-b486-223d8eb49142.png)

Wall is an image converted to instance mesh of cubes colored to match pixels.
Click or pull trigger to explode image
In VR, point at wall to highlight pixels.  When exploded, highlighted pixels are returned to wall

**Comment by IRobot1** [comment]
Updated [VR Demo](https://ng3vr.z9.web.core.windows.net/) to include hierarchical diagram using force directed graph.

![image](https://user-images.githubusercontent.com/25032599/186664053-d5d3dd58-db64-4f59-8080-f729e27e4380.png)

Network hierarchy uses ngraph graph and forcelayout to layout the diagram.  Nodes and links are rendered via NGT using position data from forcelayout.  NGT objects are gradually added to the scene to avoid stutter when in VR. 

Network diagram is standalone component with options to adjust labels, font, size, colors and animation.

Data was taken from [Highcharts network diagram sample code](https://www.highcharts.com/demo/network-graph).  I highly recommend using Highcharts for any professional charting work.  License price is reasonable and nice company to work with.

**Comment by IRobot1** [comment]
Updated [VR Demo](https://ng3vr.z9.web.core.windows.net/) to include basic 3D spirograph drawing.

![image](https://user-images.githubusercontent.com/25032599/187085343-86acce4b-cab5-485b-9b54-4d2dbf0da57a.png)

Drawing is done by rotating two arms.  One arm is attached to end of the other.  The pen tip is at the end of the second arm.  A tube mesh is generated.

GUI allows 
* adjusting rotation direction and rate of each arm
* adjust speed of redrawing mesh
* Saving mesh to either PLY or STL formats (GLTF format isn't working for some reason).  Leave VR to see downloaded files in browser.
* Model color (when not in VR)

Lots of possibilities to improve this demo.

**Comment by IRobot1** [comment]
Updated [VR Demo](https://ng3vr.z9.web.core.windows.net/) to include artificial life

![image](https://user-images.githubusercontent.com/25032599/187113593-277fe0ef-bc0a-492f-8204-e9c1b91beeb8.png)

In VR, use pointer to adjust the gravity between particle color interactions

Original idea from [this YouTube video](https://www.youtube.com/watch?v=0Kx4Y9TVMGg&t=621s)

**Comment by IRobot1** [comment]
Updated [VR Demo](https://ng3vr.z9.web.core.windows.net/) to include loading SVG icons into extruded meshes.

![image](https://user-images.githubusercontent.com/25032599/188178324-cb314ee6-f44d-44b7-b47c-1a5243e044f1.png)

Icons change every 30 seconds.  Based on data from [simple icons](https://simpleicons.org/)

**Comment by nartc** [comment]
@IRobot1 Dude, I have absolutely no words to describe my appreciation to what you've done. This is amazing. Time has been a luxury for me to tend to Angular Three and man, seeing you putting out examples using it puts a smile on my face and sets my motivation high. Thank you for doing what you do.

**Comment by IRobot1** [comment]
Updated [VR Demo](https://ng3vr.z9.web.core.windows.net/) to include directed acyclic graph example.  

![image](https://user-images.githubusercontent.com/25032599/189498511-ae6649db-269b-46e3-a9bc-4cc081fa2ca4.png)

Limited to 2D, but better layout management than forced layout.   Forced layout works best for tree structured data.
DAG is the underlying technology used by [mermaid diagraming](https://mermaid.live) and D3

**Comment by IRobot1** [comment]
Updated [VR Demo](https://ng3vr.z9.web.core.windows.net/) to include object bounded collisions example

![image](https://user-images.githubusercontent.com/25032599/190517236-b4db56bf-6daf-4ed3-86cb-c16cb21aa1cf.png)

Use left controller to pop boxes, right controller to pop spheres.

This demonstrates having collision groups where the user's hand/finger is the source of collisions within a group.  Box collisions are object bounded instead of axis align bounded for more precise overlap detection.

Dispatch events fire on both meshes when collision beings, while colliding and when collision ends.  Angular events fire on VR controller on collision begin and end.

This is the foundation for creating a UI (buttons, sliders, menus) within a three scene without the need for physics collisions.  Collision manager does not require VR, but does require something to keep calling Check Collisions method regularly.

**Comment by IRobot1** [comment]
Updated [VR Demo](https://ng3vr.z9.web.core.windows.net/) to include UI components and draggable/resizable windows

![image](https://user-images.githubusercontent.com/25032599/194783569-2a8e43a3-6f90-4bb8-baa5-a8552e1d6b51.png)

Controls include label, button, icon button, radio button, check box, toggle slider, number, text, drop-down list.
Input methods include color picker, keyboard, number pad and list.
Panels contain controls and can be dragged and resize.  They can also be collapsed to save resources and expanded when needed.

In browser view, use touch and mouse to interact with controls.  Dragging is only along x and y axes.

In VR mode, use pointer to interact with controls.  To drag, highlight title and pull trigger to grab.  Click button to expand and collapse.  

Close button is not implemented since there's no way to reopen, but you get the idea.

This is a foundation for other flat UI controls.  Multi-line text box, chips, date/time pickers, tabular grid, tabs, stepper and tree should all be possible.  Theming should also be possible and might be added in a future update.

Use vertical-layout directive to automatically stack groups or controls vertically.  No need for manual positioning.  Similarly, use horizontal-layout directive to automatically position horizontally.  Both directives support an optional margin (x for horizontal and y for vertical) to add spacing between controls.

**Reply by IRobot1** [reply]
Major performance improvement.  Now uses shared materials to improve resource usage.  Geometry and Materials can be overridden for custom look.  Drag panel (window) can now be resized.

Next up, layout improvements and adding support for table/data grid.

**Comment by IRobot1** [comment]
Updated [VR Demo](https://ng3vr.z9.web.core.windows.net/) to include some animated loading shapes

![image](https://user-images.githubusercontent.com/25032599/195410648-044ce8f0-bb16-4da2-ae59-1f3da9df0512.png)

Shapes, colors, materials and animation speeds can be customized.

**Comment by IRobot1** [comment]
Updated [VR Demo](https://ng3vr.z9.web.core.windows.net/) to include NGT version of [lil-gui](https://lil-gui.georgealways.com/)

NGT
![image](https://user-images.githubusercontent.com/25032599/195480301-d97a5197-5282-45da-8a9d-634f58fa7e4f.png) 

Original lil-gui 
![image](https://user-images.githubusercontent.com/25032599/195480880-0564dac8-5cb5-4636-afa6-a4b3f47ade9f.png)  

So far, supports basic example.  Works in browser and VR mode.

**Comment by IRobot1** [comment]
Updated [VR Demo](https://ng3vr.z9.web.core.windows.net/) to include in-scene stats panel.

![image](https://user-images.githubusercontent.com/25032599/198878891-2e8808b9-d107-4f24-abeb-9543d9119924.png)

Each specific panel can be added to a scene.  

Stats can also be placed in a UI for easy movement and interaction around the scene

Click to change between FPS, MS and MB.

**Comment by IRobot1** [comment]
Updated [VR Demo](https://ng3vr.z9.web.core.windows.net/) to include UI tab and data grid components.  Grid supports optional pagination component.  Header, footer and data cells configured using ng-templates.

![image](https://user-images.githubusercontent.com/25032599/200180722-5aabab1d-5099-42d9-81db-25a66cccf300.png)

**Comment by IRobot1** [comment]
Updated [VR Demo](https://ng3vr.z9.web.core.windows.net/) to include a basic Kanban board.  Uses new card, avatar, divider and material icon UI components and updated list and button components using ng-template.  Task cards can be dragged and dropped between columns.

![image](https://user-images.githubusercontent.com/25032599/201492654-48e59f66-0b5a-4fcf-b497-c37e6d7e5905.png)

The number of columns can be customized
The type of tasks, color and icons can be customized
The avatar icons can be customized

Would still require a lot more effort to make it fully re-usable.

**Comment by IRobot1** [comment]
Updated [VR Demo](https://ng3vr.z9.web.core.windows.net/) to include examples of [jdenticon](https://jdenticon.com/) and [DiceBear Avatars](https://avatars.dicebear.com/).  jdenticon's can be rendered to a texture or SVG geometry.  Avatars demonstrate SVG geometry only due to overlapping shapes.

![image](https://user-images.githubusercontent.com/25032599/204039468-ab1ba7b7-3db3-45f4-bfc3-cb1aeb201b9a.png)

![image](https://user-images.githubusercontent.com/25032599/204039486-1b19e2af-06e1-4ea3-9021-c63bbb77c39f.png)

**Comment by IRobot1** [comment]
Updated [VR Demo](https://ng3vr.z9.web.core.windows.net/) to include path editor that works booth in browser and VR.  This was inspired by [SvgPathEditor (yqnn.github.io)](https://yqnn.github.io/svg-path-editor/)

![image](https://user-images.githubusercontent.com/25032599/208776644-6b4d52a5-ed13-4f28-bcc7-97a9d0799b2a.png)

Uses new mini menu UI component for choosing next curve type. Menu buttons support either text label or material icon.
![image](https://user-images.githubusercontent.com/25032599/208776898-dcbae1f1-5166-489a-afb0-c65beaa533f2.png)

The following keyboard shorts are supported
* L - add line
* V - add vertical line
* H - add horizontal line
* C - add cubic curve (2 control points)
* Q - add Bezier curve (1 control point)
* Del - delete current path command

There's options to show/hide the grid and points, preview as extruded and lathed geometry.

**Comment by IRobot1** [comment]
Updated [VR Demo](https://ng3vr.z9.web.core.windows.net/) to include examples of using Pexels API to search and show photos and videos.

![image](https://user-images.githubusercontent.com/25032599/209473534-83b1d49e-3c16-4ba1-b618-160f01582da8.png)

UI supports entry of key before API can be used.  Use the paste icon to paste key and Enter to continue. Key is saved as a cookie.
![image](https://user-images.githubusercontent.com/25032599/209474227-621c5391-da93-459e-919d-2f7eb5d1d0f6.png)


Works on desktop and in VR.  Sound works on video.  Controls to mute and loop a video.

Using this as a foundation, should be possible to create image galleries (carousel) or video media player with controls and sound effects.

**Reply by IRobot1** [reply]
Updated [VR Demo](https://ng3vr.z9.web.core.windows.net/) to include charts.  Pan around and zoom in for closer look.
![image](https://user-images.githubusercontent.com/25032599/216835584-fa621b6d-6831-4489-ac9c-700f110cf26f.png)

Bar charts using Angular templates allows customizing axis text and labels above bars.  Chart supports custom geometry for each bar.
Stacked bar charts support flat or pointed top and shape
Line plot supports showing markers and live value updates
Area plot supports live value updates
Pie chart supports adding space between segments
Components for grid and simple axis presentation. Both support custom width and height.

All charts allow customizing materials/colors.  None are interactive.

The following image was used as inspiration for creating these basic chart components
![image](https://user-images.githubusercontent.com/25032599/216835865-3e1c1be5-f080-4428-b6e8-5b1af3a6e5f0.png)

**Comment by nartc** [comment]
Hi @IRobot1, are you good with recording videos on the VR stuffs as maybe not a lot of people have VR set to check out these examples?

**Reply by IRobot1** [reply]
Most don't require a VR headset to full use.  However, they're best experience is with a headset. I'll make it clearer which one's required a headset.

As for recording videos, I haven't done this for years.  Its on my to do list.  This may all change with your new version. so I'll holding off until all the examples and project are migrated to the new version.

I was planning on doing some walk throughs on re-creating some of the examples from scratch and cover some does and don'ts.   

In the mean time, if you want to mention the examples in a future video, that might be a good place to start.  Its a huge body of work with lots of working examples and variety of technologies in one place.
